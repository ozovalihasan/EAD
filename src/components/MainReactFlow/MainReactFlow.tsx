import { useState, useRef, useCallback, memo } from 'react';
import ReactFlow, {
  ReactFlowProvider,
  Node,
  ReactFlowInstance,
  Controls
} from 'react-flow-renderer';

import useStore from '@/zustandStore/store';

import {
  EntityNode,
  EntityNodeDataType,
  HasManyEdge,
  HasOneEdge,
  ThroughEdge,
  ConnectionLine,

} from "@/components"

const nodeTypes = {
  entity: EntityNode,
};
const edgeTypes = {
  hasMany: HasManyEdge,
  hasOne: HasOneEdge,
  through: ThroughEdge,
};

export const MainReactFlow = memo(() => {

  const nodes = useStore(store => store.nodes)
  const edges = useStore(store => store.edges)
  const onNodesChange = useStore(store => store.onNodesChange)
  const onEdgesChange = useStore(store => store.onEdgesChange)
  const onConnect = useStore(store => store.onConnect)
  const addNode = useStore(store => store.addNode)
  const onConnectStart = useStore(store => store.onConnectStart)
  const onConnectEnd = useStore(store => store.onConnectEnd)
  const onNodeMouseEnter = useStore(store => store.onNodeMouseEnter)
  const onNodeMouseLeave = useStore(store => store.onNodeMouseLeave)
  const onEdgeMouseEnter = useStore(store => store.onEdgeMouseEnter)
  const onEdgeMouseLeave = useStore(store => store.onEdgeMouseLeave)


  const reactFlowWrapper = useRef<HTMLDivElement>(null);
  const [reactFlowInstance, setReactFlowInstance] = useState<ReactFlowInstance | null>(null);


  const onDragOver = useCallback((event: React.DragEvent<HTMLDivElement>) => {
    event.preventDefault();
    event.dataTransfer.dropEffect = 'move';
  }, []);

  const onDrop = useCallback(
    (event: React.DragEvent<HTMLDivElement>) => {
      event.preventDefault();

      const reactFlowBounds: DOMRect = (reactFlowWrapper.current!).getBoundingClientRect();
      
      const type = event.dataTransfer.getData('application/reactflow');
      const tableId = event.dataTransfer.getData('tableId');

      if (typeof type === 'undefined' || !type) {
        return;
      }

      const position: {x: number, y: number} = (reactFlowInstance!).project({
          x: event.clientX - reactFlowBounds.left,
          y: event.clientY - reactFlowBounds.top,
        });

      
      const name = useStore.getState().tables[tableId].name;

      const newNode: Node<EntityNodeDataType> = {
        id: useStore.getState().idCounter.toString(),
        type: "entity",
        position,
        data: { tableId, name },
      };
      
      useStore.getState().increaseIdCounter()
      
      addNode(newNode);
    },
    [reactFlowInstance]
  );
  
  return (
    <div className="h-full flex-grow relative" ref={reactFlowWrapper}>
      <ReactFlowProvider>      
        <ReactFlow
          nodes={nodes}
          edges={edges}
          onNodesChange={onNodesChange}
          onEdgesChange={onEdgesChange}
          onConnect={onConnect}
          onInit={setReactFlowInstance}
          onDrop={onDrop}
          onConnectStart={onConnectStart}
          onConnectEnd={onConnectEnd}
          onNodeMouseEnter={onNodeMouseEnter}
          onNodeMouseLeave={onNodeMouseLeave}
          onEdgeMouseEnter={onEdgeMouseEnter}
          onEdgeMouseLeave={onEdgeMouseLeave}
          onDragOver={onDragOver}
          edgeTypes={edgeTypes}
          deleteKeyCode={"Delete"}
          connectionLineComponent={ConnectionLine }
          elevateEdgesOnSelect={true}
          fitView
          nodeTypes={nodeTypes}
          snapToGrid={true}
        >
          <Controls />
          
        </ReactFlow>
      </ReactFlowProvider>
      
    </div>
  )
})
