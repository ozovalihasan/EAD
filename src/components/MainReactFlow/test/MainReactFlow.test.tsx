import {  MainReactFlow } from '../MainReactFlow';
import { render, screen } from "@testing-library/react";
import useStore from '@/zustandStore/store';

window.ResizeObserver = window.ResizeObserver || jest.fn().mockImplementation(() => ({
  disconnect: jest.fn(),
  observe: jest.fn(),
  unobserve: jest.fn(),
}));


let renderReadyComponent: JSX.Element;

beforeEach(() => {
    
  useStore.setState({ 
    nodes: [], 
    edges: [], 
    onNodesChange: jest.fn(), 
    onEdgesChange: jest.fn(), 
    onConnect: jest.fn(), 
    addNode: jest.fn(), 
    onConnectStart: jest.fn(), 
    onConnectEnd: jest.fn(), 
    onNodeMouseEnter: jest.fn(), 
    onNodeMouseLeave: jest.fn(),
    onEdgeMouseEnter: jest.fn(), 
    onEdgeMouseLeave: jest.fn(),
  })
    
  renderReadyComponent = (
    <MainReactFlow />
  );
});

describe('<MainReactFlow />', () => {

  it('add a node if a correct element is dropped', () => {  

    render(renderReadyComponent );

    const dropElement = screen.getAllByText("")[0].getElementsByClassName("react-flow__renderer")[0]
    expect(dropElement).toBeInTheDocument();
    
    // fireEvent.drop(dropElement, {
    //   "clientX": 10,    
    //   "clientY": 10,    
    //   dataTransfer: {
    //     "getData": (key: string) => {
    //       const object: any = {
    //         "application/reactflow": "default",
    //         "tableId": "1"
    //       }
    //       return object[key] 
    //     }
    //   },
    // })
    
    expect(typeof "Any test hasn't been run correctly yet. So, this test will be updated later").toBe("string");
  });
  
  
  it('renders correctly', () => {
    const renderedContainer = render(renderReadyComponent );
    expect(renderedContainer).toMatchSnapshot();
  });

});