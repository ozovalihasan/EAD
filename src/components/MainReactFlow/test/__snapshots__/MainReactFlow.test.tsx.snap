// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainReactFlow /> renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="h-full flex-grow relative"
      >
        <div
          class="react-flow light"
          data-testid="rf__wrapper"
          style="width: 100%; height: 100%; overflow: hidden; position: relative; z-index: 0;"
        >
          <div
            class="react-flow__renderer"
            style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px;"
          >
            <div
              class="react-flow__pane draggable"
              style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px;"
            >
              <div
                class="react-flow__viewport xyflow__viewport react-flow__container"
                style="transform: translate(0px,0px) scale(1);"
              >
                <div
                  class="react-flow__edges"
                />
                <div
                  class="react-flow__edgelabel-renderer"
                />
                <div
                  class="react-flow__nodes"
                  style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px;"
                />
                <div
                  class="react-flow__viewport-portal"
                />
              </div>
            </div>
          </div>
          <div
            aria-label="React Flow controls"
            class="react-flow__panel react-flow__controls vertical bottom left"
            data-testid="rf__controls"
            style="pointer-events: all;"
          >
            <button
              aria-label="zoom in"
              class="react-flow__controls-button react-flow__controls-zoomin"
              title="zoom in"
              type="button"
            >
              <svg
                viewBox="0 0 32 32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M32 18.133H18.133V32h-4.266V18.133H0v-4.266h13.867V0h4.266v13.867H32z"
                />
              </svg>
            </button>
            <button
              aria-label="zoom out"
              class="react-flow__controls-button react-flow__controls-zoomout"
              title="zoom out"
              type="button"
            >
              <svg
                viewBox="0 0 32 5"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h32v4.2H0z"
                />
              </svg>
            </button>
            <button
              aria-label="fit view"
              class="react-flow__controls-button react-flow__controls-fitview"
              title="fit view"
              type="button"
            >
              <svg
                viewBox="0 0 32 30"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M3.692 4.63c0-.53.4-.938.939-.938h5.215V0H4.708C2.13 0 0 2.054 0 4.63v5.216h3.692V4.631zM27.354 0h-5.2v3.692h5.17c.53 0 .984.4.984.939v5.215H32V4.631A4.624 4.624 0 0027.354 0zm.954 24.83c0 .532-.4.94-.939.94h-5.215v3.768h5.215c2.577 0 4.631-2.13 4.631-4.707v-5.139h-3.692v5.139zm-23.677.94c-.531 0-.939-.4-.939-.94v-5.138H0v5.139c0 2.577 2.13 4.707 4.708 4.707h5.138V25.77H4.631z"
                />
              </svg>
            </button>
            <button
              aria-label="toggle interactivity"
              class="react-flow__controls-button react-flow__controls-interactive"
              title="toggle interactivity"
              type="button"
            >
              <svg
                viewBox="0 0 25 32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M21.333 10.667H19.81V7.619C19.81 3.429 16.38 0 12.19 0c-4.114 1.828-1.37 2.133.305 2.438 1.676.305 4.42 2.59 4.42 5.181v3.048H3.047A3.056 3.056 0 000 13.714v15.238A3.056 3.056 0 003.048 32h18.285a3.056 3.056 0 003.048-3.048V13.714a3.056 3.056 0 00-3.048-3.047zM12.19 24.533a3.056 3.056 0 01-3.047-3.047 3.056 3.056 0 013.047-3.048 3.056 3.056 0 013.048 3.048 3.056 3.056 0 01-3.048 3.047z"
                />
              </svg>
            </button>
          </div>
          <div
            class="absolute right-10 bottom-10 z-10"
          >
            MockRemoveEdgeButton
          </div>
          <div
            class="react-flow__panel react-flow__attribution bottom right"
            data-message="Please only hide this attribution when you are subscribed to React Flow Pro: https://pro.reactflow.dev"
            style="pointer-events: all;"
          >
            <a
              aria-label="React Flow attribution"
              href="https://reactflow.dev"
              rel="noopener noreferrer"
              target="_blank"
            >
              React Flow
            </a>
          </div>
          <div
            id="react-flow__node-desc-1"
            style="display: none;"
          >
            Press enter or space to select a node.
            You can then use the arrow keys to move the node around.
             Press delete to remove it and escape to cancel.
             
          </div>
          <div
            id="react-flow__edge-desc-1"
            style="display: none;"
          >
            Press enter or space to select an edge. You can then press delete to remove it or escape to cancel.
          </div>
          <div
            aria-atomic="true"
            aria-live="assertive"
            id="react-flow__aria-live-1"
            style="position: absolute; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); clip-path: inset(100%);"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="h-full flex-grow relative"
    >
      <div
        class="react-flow light"
        data-testid="rf__wrapper"
        style="width: 100%; height: 100%; overflow: hidden; position: relative; z-index: 0;"
      >
        <div
          class="react-flow__renderer"
          style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px;"
        >
          <div
            class="react-flow__pane draggable"
            style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px;"
          >
            <div
              class="react-flow__viewport xyflow__viewport react-flow__container"
              style="transform: translate(0px,0px) scale(1);"
            >
              <div
                class="react-flow__edges"
              />
              <div
                class="react-flow__edgelabel-renderer"
              />
              <div
                class="react-flow__nodes"
                style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px;"
              />
              <div
                class="react-flow__viewport-portal"
              />
            </div>
          </div>
        </div>
        <div
          aria-label="React Flow controls"
          class="react-flow__panel react-flow__controls vertical bottom left"
          data-testid="rf__controls"
          style="pointer-events: all;"
        >
          <button
            aria-label="zoom in"
            class="react-flow__controls-button react-flow__controls-zoomin"
            title="zoom in"
            type="button"
          >
            <svg
              viewBox="0 0 32 32"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M32 18.133H18.133V32h-4.266V18.133H0v-4.266h13.867V0h4.266v13.867H32z"
              />
            </svg>
          </button>
          <button
            aria-label="zoom out"
            class="react-flow__controls-button react-flow__controls-zoomout"
            title="zoom out"
            type="button"
          >
            <svg
              viewBox="0 0 32 5"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h32v4.2H0z"
              />
            </svg>
          </button>
          <button
            aria-label="fit view"
            class="react-flow__controls-button react-flow__controls-fitview"
            title="fit view"
            type="button"
          >
            <svg
              viewBox="0 0 32 30"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3.692 4.63c0-.53.4-.938.939-.938h5.215V0H4.708C2.13 0 0 2.054 0 4.63v5.216h3.692V4.631zM27.354 0h-5.2v3.692h5.17c.53 0 .984.4.984.939v5.215H32V4.631A4.624 4.624 0 0027.354 0zm.954 24.83c0 .532-.4.94-.939.94h-5.215v3.768h5.215c2.577 0 4.631-2.13 4.631-4.707v-5.139h-3.692v5.139zm-23.677.94c-.531 0-.939-.4-.939-.94v-5.138H0v5.139c0 2.577 2.13 4.707 4.708 4.707h5.138V25.77H4.631z"
              />
            </svg>
          </button>
          <button
            aria-label="toggle interactivity"
            class="react-flow__controls-button react-flow__controls-interactive"
            title="toggle interactivity"
            type="button"
          >
            <svg
              viewBox="0 0 25 32"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21.333 10.667H19.81V7.619C19.81 3.429 16.38 0 12.19 0c-4.114 1.828-1.37 2.133.305 2.438 1.676.305 4.42 2.59 4.42 5.181v3.048H3.047A3.056 3.056 0 000 13.714v15.238A3.056 3.056 0 003.048 32h18.285a3.056 3.056 0 003.048-3.048V13.714a3.056 3.056 0 00-3.048-3.047zM12.19 24.533a3.056 3.056 0 01-3.047-3.047 3.056 3.056 0 013.047-3.048 3.056 3.056 0 013.048 3.048 3.056 3.056 0 01-3.048 3.047z"
              />
            </svg>
          </button>
        </div>
        <div
          class="absolute right-10 bottom-10 z-10"
        >
          MockRemoveEdgeButton
        </div>
        <div
          class="react-flow__panel react-flow__attribution bottom right"
          data-message="Please only hide this attribution when you are subscribed to React Flow Pro: https://pro.reactflow.dev"
          style="pointer-events: all;"
        >
          <a
            aria-label="React Flow attribution"
            href="https://reactflow.dev"
            rel="noopener noreferrer"
            target="_blank"
          >
            React Flow
          </a>
        </div>
        <div
          id="react-flow__node-desc-1"
          style="display: none;"
        >
          Press enter or space to select a node.
          You can then use the arrow keys to move the node around.
           Press delete to remove it and escape to cancel.
           
        </div>
        <div
          id="react-flow__edge-desc-1"
          style="display: none;"
        >
          Press enter or space to select an edge. You can then press delete to remove it or escape to cancel.
        </div>
        <div
          aria-atomic="true"
          aria-live="assertive"
          id="react-flow__aria-live-1"
          style="position: absolute; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); clip-path: inset(100%);"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
